js-pkg/client/src/indexeddb.ts

```diff
@@ -0,0 +1,118 @@
+import { Doc } from 'yjs'
+import * as Y from 'yjs'
+import { getOrCreateKey } from './keystore'
+import { decryptData, encryptData } from './encryption'
+
+const DB_PREFIX = 'y-sweet-'
+const OBJECT_STORE_NAME = 'updates'
+
+/** Maximum number of independent updates to store in IndexedDB.
+ * If this is exceeded, all of the updates are compacted into one.
+ */
+const MAX_UPDATES_IN_STORE = 10
+
+function openIndexedDB(name: string): Promise<IDBDatabase> {
+  return new Promise<IDBDatabase>((resolve, reject) => {
+    const request = indexedDB.open(name, 2)
+    request.onupgradeneeded = (event: IDBVersionChangeEvent) => {
+      const db = (event.target! as any).result as IDBDatabase
+      db.createObjectStore(OBJECT_STORE_NAME, { autoIncrement: true })
+    }
+    request.onsuccess = () => resolve(request.result)
+    request.onerror = reject
+  })
+}
+
+export class IndexedDBProvider {
+  db: Promise<IDBDatabase>
+  objectCount: number = 0
+  encryptionKey: Promise<CryptoKey>
```

Comment:
Not sure how to test `await` rather than `then` but if you run this function it prints "no promise" first, and then "promise"

```js
function test() {
  Promise.resolve().then(() => console.log("promise"))
  console.log("no promise")
}
```